db_util.go

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2022 Red Hat, Inc.
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ */
+
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	c "nav/constants"
+)
+
+type Datasource interface {
+	init(arg interface{}) (err error)
+	GetExploredSubsystemByName(subs string) string
+	getSuccessorsById(symbolId int, instance int) ([]entry, error)
+	getSubsysFromSymbolName(symbol string, instance int) (string, error)
+	sym2num(symb string, instance int) (int, error)
+	symbSubsys(symblist []int, instance int) (string, error)
+	getEntryById(symbolId int, instance int) (entry, error)
+}
+
+const SUBSYS_UNDEF = "The REST"
+
+// Parent node.
+type node struct {
+	subsys     string
+	symbol     string
+	sourceRef  string
+	addressRef string
+}
+type adjM struct {
+	l node
+	r node
+}
+
+type entry struct {
+	symbol     string
+	fn         string
+	sourceRef  string
+	addressRef string
+	subsys     []string
+	symId      int
+}
+
+type edge struct {
+	sourceRef  string
+	addressRef string
+	caller     int
+	callee     int
+}
+
+// Return id an item is already in the list.
+func notIn(list []int, v int) bool {
+
+	for _, a := range list {
+		if a == v {
+			return false
+		}
+	}
+	return true
+}
+
+// Removes duplicates resulting by the exploration of a call tree.
+func removeDuplicate(list []entry) []entry {
+
+	sort.SliceStable(list, func(i, j int) bool { return list[i].symId < list[j].symId })
+	allKeys := make(map[int]bool)
+	var res []entry
+	for _, item := range list {
+		if _, value := allKeys[item.symId]; !value {
+			allKeys[item.symId] = true
+			res = append(res, item)
+		}
+	}
+	return res
+}
+
+// Checks if a given function needs to be explored.
+func notExcluded(symbol string, excluded []string) bool {
+	for _, s := range excluded {
+		if match, _ := regexp.MatchString(s, symbol); match {
+			return false
+		}
+	}
+	return true
+}
+
+// Computes the call tree of a given function name
+// TODO: refactory needed:
+// What is the problem: too many args.
+// suggestion: New version with input and output structs.
+func navigate(d Datasource, symbolId int, parentDispaly node, targets []string, visited *[]int, AdjMap *[]adjM, prod map[string]int, instance int, mode c.OutMode, excludedAfter []string, excludedBefore []string, depth int, maxdepth int, dotFmt string, output *string) {
+	var tmp, s string
+	var l, r, ll node
+
+	*visited = append(*visited, symbolId)
+	l = parentDispaly
+	successors, err := d.getSuccessorsById(symbolId, instance)
+	if mode == c.PrintAll {
+		successors = removeDuplicate(successors)
+	}
+	if err == nil {
```

Comment:
Yes, I agree with you. The current form of the function is a result of previous versions and may not be optimal. Upon reflection, it makes no sense to continue the computation if an error is produced. Therefore, it would be beneficial to modify this function to report the error condition externally.